rfcs/0002-validation/text.md

```diff
@@ -0,0 +1,149 @@
+2. Artifact bucket. The S3 bucket where artifacts associated with datasets are stored and served.
+3. cellxgene deployment bucket. The S3 bucket configured to serve datasets via cellxgene.
+4. Portal database. The database where information about datasets is stored. The portal API queries this database when
+   responding to requests.
+5. Validation and processing image. The Docker image containing the dependencies and script to validate and process.
+6. Validation and processing Fargate task. The task definition for executing the image with AWS Fargate.
+
+#### Portal Database Schema
+
+The relevant part of the portal database is three tables: `dataset`, `artifacts`, and `deployment_directories`.
+
+![Database Schema](imgs/db_schema.png)
+
+The `dataset` table tracks the status of datasets and various metadata about them that is used by the portal frontend.
+There are three status fields:
+
+- upload\_status: uploading, uploaded, canceled
+- validation\_status: validating, valid, invalid
+- artifact\_status: creating, complete
+
+The `artifacts` table tracks the different downloadable files associated with a dataset. This stores the format of the
+download and the bucket and object key in the artifact bucket.
+
+The `deployment_directories` table tracks the cellxgene deployments associated with a dataset and the URL where the
+deployment can be accessed.
+
+### Handoff from Upload
+
+At the end of the upload process there must be two elements present in AWS:
+
+1. A row in the `dataset` table with `id` and `collection_id` populated and `upload_status` set to "uploaded".
+2. A dataset file in the staging bucket with a `x-amz-meta-dataset-id` metadata field whose value is the `id` of the new
+   row in the `dataset` table.
+
+A [CloudWatch Event](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/CloudWatch-Events-tutorial-ECS.html)
+will trigger the execution of the Fargate task when a new dataset file is added to the staging bucket.
+
+### Steps of the Fargate Task
+
+1. Check the `dataset` table to verify that the expected row exists and `upload_status` is "uploaded". To avoid possible
+   race conditions from the upload process, wait and recheck if this isn't true.
+2. Set the `validation_status` in `dataset` to "validating".
+3. Localize the uploaded file. HDF5 doesn't usually work with remote files very well.
+4. Run the `cellxgene schema validate` command to do a shallow validation of the dataset file.
+   - If this has a non-zero exit code, set `validation_status` to "invalid", write stdout from the command to
+     `processing_message`, and halt.
+   - Otherwise, set `validation_status` to "valid" and `artifact_status` to "creating".
+5. Read the appropriate values from the dataset file and upload the JSONB fields in the `dataset` table for tissue,
+   assay, etc.
+6. For format in `{loom, h5ad, rds}`
+   - If the format does not match the format of the uploaded dataset file, run a conversion script.
```

Comment:
Is it ever possible for a validated .h5ad dataset to fail conversion and require intervention?

Reply:
I see this is touched on [Monitoring and Error Reporting](https://github.com/chanzuckerberg/single-cell/blob/c0bef7684a0cd1f523ff3632918ec4b4a273863f/rfcs/0002-validation/text.md#monitoring-and-error-reporting) as _unexpected failure_?  ... but I'm curious about how this is reported and how it may impact the related UX experience  (for example, there's no artifact failure state?)

@hthomas-czi - calling to your attention too.

Reply:
Yeah it's definitely possible. I mean it means there's a bug of some kind, but tbh the conversion process is relatively buggy.

I think if there's a failure, we aim to update the status in the DB so the frontend can at least tell the user something went wrong, and then we get an alert so we can dig in an debug.

Reply:
I guess the question is whether the UX would share partial results or not if other format conversions were successful and the CXG were generated. (For example, only the .rds dataset conversion failed).

Reply:
RE _we aim to update the status in the DB_ - I think that there's a gap in the current status values per my earlier comment.

Reply:
It'd be a better experience if we showed partial results when possible. It's entirely plausible that the user doesn't care about the format that failed conversion. We'd just need to reflect that in the UI and provide a recommendation. I [mocked-up an idea](https://www.figma.com/file/g0L0CfVzRc8r6HQa4NH1wJ/Collections?node-id=174%3A0) for how we could do that.